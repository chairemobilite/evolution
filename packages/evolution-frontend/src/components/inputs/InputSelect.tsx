/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { withTranslation, WithTranslation } from 'react-i18next';

import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import { InputSelectType, isGroupedChoice, ChoiceType } from '../../services/widgets';
import { UserInterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import * as surveyHelper from '../../utils/helpers';
import { FrontendUser } from 'chaire-lib-frontend/lib/services/auth/user';

export interface InputSelectProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> {
    id: string;
    onValueChange?: (e: any) => void;
    value?: string;
    interview: UserInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
    // TODO There's also a path in widgetConfig, but this one comes from the props of the question. See if it's always the same and use the one from widgetConfig if necessary
    path: string;
    user: FrontendUser;
    inputRef?: React.LegacyRef<HTMLInputElement>;
    size?: 'small' | 'medium' | 'large';
    widgetConfig: InputSelectType<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
}

export interface InputSelectOptionProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> {
    choice: ChoiceType<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
    interview: UserInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
    // TODO There's also a path in widgetConfig, but this one comes from the props of the question. See if it's always the same and use the one from widgetConfig if necessary
    path: string;
    user: FrontendUser;
}

const InputSelectOption = <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    props: InputSelectOptionProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> & WithTranslation
) => {
    if (
        props.choice.hidden === true ||
        !surveyHelper.parseBoolean(props.choice.conditional, props.interview, props.path, props.user)
    ) {
        return null;
    }
    return (
        <option
            key={`input-select-container__${props.choice.value}`}
            value={props.choice.value}
            className={'input-select-option'}
        >
            {surveyHelper.parseString(
                props.choice.label[props.i18n.language] || props.choice.label,
                props.interview,
                props.path,
                props.user
            )}
        </option>
    );
};

const InputSelectOptionT = withTranslation()(InputSelectOption) as React.FunctionComponent<
    InputSelectOptionProps<any, any, any, any>
>;

export const InputSelect = <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    props: InputSelectProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> & WithTranslation
) => {
    const widgetChoices = props.widgetConfig.choices;
    const choices =
        typeof widgetChoices === 'function' ? widgetChoices(props.interview, props.path, props.user) : widgetChoices;

    let isGrouped = false;
    const selectOptions = choices.map((choice) => {
        if (isGroupedChoice(choice)) {
            isGrouped = true;
            const groupSelectOptions = choice.choices.map((groupedChoice, index) => (
                <InputSelectOptionT
                    key={`${props.id}_opt_${index}`}
                    choice={groupedChoice as any}
                    path={props.path}
                    interview={props.interview}
                    user={props.user}
                />
            ));

            return (
                <optgroup
                    key={`input-select-group-container__${choice.groupShortname}`}
                    label={choice.groupLabel[props.i18n.language] || choice.groupLabel}
                >
                    {groupSelectOptions}
                </optgroup>
            );
        } else {
            return (
                <InputSelectOptionT
                    key={`${props.id}_opt_${choice.value}`}
                    choice={choice as any}
                    path={props.path}
                    interview={props.interview}
                    user={props.user}
                />
            );
        }
    }, this);

    return (
        <div className="survey-question__input-select-container">
            <select
                id={props.id}
                onChange={props.onValueChange}
                value={_isBlank(props.value) ? '' : props.value}
                style={{
                    overflow: 'hidden',
                    whiteSpace: 'nowrap',
                    textOverflow: 'ellipsis',
                    maxWidth: '100%'
                }}
            >
                <option key={'input-select-container__blank'} value="" className={'input-select-option'}></option>

                {selectOptions}
                {
                    isGrouped === false && (
                        <optgroup key="-1" label=""></optgroup>
                    ) /* if no group, add an empty group (hack) to make sure <option> with long text are not truncated on small width screens (iphone) */
                }
            </select>
        </div>
    );
};

export default withTranslation()(InputSelect) as React.FunctionComponent<InputSelectProps<any, any, any, any>>;
