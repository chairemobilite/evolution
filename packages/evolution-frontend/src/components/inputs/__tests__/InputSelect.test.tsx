/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import TestRenderer from 'react-test-renderer';

import { interviewAttributes } from './interviewData.test';
import InputSelect from '../InputSelect';
import i18next from 'i18next';

const userAttributes = {
    id: 1,
    username: 'foo',
    preferences: {  },
    serializedPermissions: [],
    isAuthorized: () => true,
    is_admin: false,
    pages: [],
    showUserInfo: true
}

test('Render InputSelect with normal option type', () => {

    const conditionalFct = jest.fn().mockReturnValue(true);
    const translationFct = jest.fn().mockReturnValue('Translated string');
    const choices = [
        {
            value: 'val1',
            label: { en: 'english value', fr: 'valeur française' },
            hidden: false,
            icon: 'creative-commons' as const
        },
        {
            value: 'val2',
            label: 'Unilingual label',
            iconPath: 'img/test.png'
        },
        {
            value: 'hiddenVal',
            label: { en: 'english hidden', fr: 'cachée' },
            hidden: true
        },
        {
            value: 'conditionalVal',
            label: { en: 'english conditional', fr: 'conditionnelle' },
            conditional: conditionalFct
        },
        {
            value: 'val3',
            label: translationFct,
            hidden: false,
            color: 'green'
        }
    ];

    const widgetConfig = {
        type: 'question' as const,
        twoColumns: true,
        path: 'test.foo',
        inputType: 'select' as const,
        choices,
        size: 'medium',
        containsHtml: true,
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    };
    
    const wrapper = TestRenderer.create(
        <InputSelect
            id={'test'}
            onValueChange={() => { /* nothing to do */}}
            widgetConfig={widgetConfig}
            value='value'
            inputRef={React.createRef()}
            interview={interviewAttributes}
            user={userAttributes}
            path='foo.test'
        />
    );
    expect(wrapper).toMatchSnapshot();
    expect(conditionalFct).toHaveBeenCalledTimes(1);
    expect(conditionalFct).toHaveBeenCalledWith(interviewAttributes, 'foo.test', userAttributes);
    expect(translationFct).toHaveBeenCalledWith(i18next.t, interviewAttributes, 'foo.test', userAttributes);
    
});

test('Render InputSelect with choice function and grouped choice type', () => {

    const choices = [
        {
            value: 'val1',
            label: { en: 'english value', fr: 'valeur française' },
            hidden: false,
            icon: 'creative-commons' as const
        },
        {
            groupName: 'select group',
            groupShortname: 'sel',
            groupLabel: { en: 'english select label', fr: 'étiquette française' },
            choices: [
                {
                    value: 'val2',
                    label: 'Unilingual label',
                    iconPath: 'img/test.png'
                },
                {
                    value: 'hiddenVal',
                    label: { en: 'english hidden', fr: 'cachée' },
                    hidden: true
                },
            ]
        },
        {
            groupName: 'select group',
            groupShortname: 'sel',
            groupLabel: 'Unilingual group label',
            choices: [
                {
                    value: 'val2',
                    label: 'Unilingual label',
                    iconPath: 'img/test.png'
                },
                {
                    value: 'val3',
                    label: { en: 'english val3', fr: 'val3 français' }
                },
            ]
        }
    ];
    const choiceFct = jest.fn().mockReturnValue(choices);
    
    const widgetConfig = {
        type: 'question' as const,
        twoColumns: true,
        path: 'test.foo',
        inputType: 'select' as const,
        choices: choiceFct,
        containsHtml: true,
        size: 'medium',
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    };
    
    const wrapper = TestRenderer.create(
        <InputSelect
            id={'test'}
            onValueChange={() => { /* nothing to do */}}
            widgetConfig={widgetConfig}
            value='value'
            inputRef={React.createRef()}
            interview={interviewAttributes}
            user={userAttributes}
            path='foo.test'
        />
    );
    expect(wrapper).toMatchSnapshot();
    expect(choiceFct).toHaveBeenCalledTimes(1);
    expect(choiceFct).toHaveBeenCalledWith(interviewAttributes, 'foo.test', userAttributes);
    
});

test('Render InputSelect with grouped choice type without labels', () => {

    const choices = [
        {
            value: 'val1',
            label: { en: 'english value', fr: 'valeur française' },
            hidden: false,
            icon: 'creative-commons' as const
        },
        {
            groupName: 'select group',
            groupShortname: 'sel',
            groupLabel: '',
            choices: [
                {
                    value: 'val2',
                    label: 'Unilingual label',
                    iconPath: 'img/test.png'
                },
                {
                    value: 'hiddenVal',
                    label: { en: 'english hidden', fr: 'cachée' },
                    hidden: true
                },
            ]
        },
        {
            groupName: 'select group',
            groupShortname: 'sel',
            groupLabel: '',
            choices: [
                {
                    value: 'val2',
                    label: 'Unilingual label',
                    iconPath: 'img/test.png'
                },
                {
                    value: 'val3',
                    label: { en: 'english val3', fr: 'val3 français' }
                },
            ]
        }
    ];

    const widgetConfig = {
        type: 'question' as const,
        twoColumns: true,
        path: 'test.foo',
        inputType: 'select' as const,
        choices,
        containsHtml: true,
        size: 'medium',
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    };

    const wrapper = TestRenderer.create(
        <InputSelect
            id={'test'}
            onValueChange={() => { /* nothing to do */}}
            widgetConfig={widgetConfig}
            value='value'
            inputRef={React.createRef()}
            interview={interviewAttributes}
            user={userAttributes}
            path='foo.test'
        />
    );
    expect(wrapper).toMatchSnapshot();

});
