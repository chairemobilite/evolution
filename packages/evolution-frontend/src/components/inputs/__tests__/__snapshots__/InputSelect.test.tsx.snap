// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render InputSelect with choice function and grouped choice type 1`] = `
<div
  className="survey-question__input-select-container"
>
  <select
    id="test"
    onChange={[Function]}
    style={
      Object {
        "maxWidth": "100%",
        "overflow": "hidden",
        "textOverflow": "ellipsis",
        "whiteSpace": "nowrap",
      }
    }
    value="value"
  >
    <option
      className="input-select-option"
      value=""
    />
    <option
      className="input-select-option"
      value="val1"
    >
      english value
    </option>
    <optgroup
      label="english select label"
    >
      <option
        className="input-select-option"
        value="val2"
      >
        Unilingual label
      </option>
    </optgroup>
    <optgroup
      label="Unilingual group label"
    >
      <option
        className="input-select-option"
        value="val2"
      >
        Unilingual label
      </option>
      <option
        className="input-select-option"
        value="val3"
      >
        english val3
      </option>
    </optgroup>
  </select>
</div>
`;

exports[`Render InputSelect with grouped choice type without labels 1`] = `
<div
  className="survey-question__input-select-container"
>
  <select
    id="test"
    onChange={[Function]}
    style={
      Object {
        "maxWidth": "100%",
        "overflow": "hidden",
        "textOverflow": "ellipsis",
        "whiteSpace": "nowrap",
      }
    }
    value="value"
  >
    <option
      className="input-select-option"
      value=""
    />
    <option
      className="input-select-option"
      value="val1"
    >
      english value
    </option>
    <option
      disabled={true}
    >
      _________
    </option>
    <option
      className="input-select-option"
      value="val2"
    >
      Unilingual label
    </option>
    <option
      disabled={true}
    >
      _________
    </option>
    <option
      className="input-select-option"
      value="val2"
    >
      Unilingual label
    </option>
    <option
      className="input-select-option"
      value="val3"
    >
      english val3
    </option>
  </select>
</div>
`;

exports[`Render InputSelect with normal option type 1`] = `
<div
  className="survey-question__input-select-container"
>
  <select
    id="test"
    onChange={[Function]}
    style={
      Object {
        "maxWidth": "100%",
        "overflow": "hidden",
        "textOverflow": "ellipsis",
        "whiteSpace": "nowrap",
      }
    }
    value="value"
  >
    <option
      className="input-select-option"
      value=""
    />
    <option
      className="input-select-option"
      value="val1"
    >
      english value
    </option>
    <option
      className="input-select-option"
      value="val2"
    >
      Unilingual label
    </option>
    <option
      className="input-select-option"
      value="conditionalVal"
    >
      english conditional
    </option>
    <option
      className="input-select-option"
      value="val3"
    >
      Translated string
    </option>
    <optgroup
      label=""
    />
  </select>
</div>
`;
